doc/BMIconventions.md

```diff
@@ -0,0 +1,313 @@
+
+## Layout and Flattening
+
+### C ordering required
+
+The [BMI best practices](https://bmi.readthedocs.io/en/stable/bmi.best_practices.html) document states:
+
+> BMI functions always use flattened, one-dimensional arrays...It’s the developer’s responsibility to ensure that array information is flattened/redimensionalized in the correct order.
+
+However, strictly speaking, *how* a multi-dimensional array should be flattened into a one-dimensional one is never directly addressed. **For ngen, it is required that flattening happens as if the array was a C array, sometimes referred to as "row major order".** That is, if you create a multi-dimensional array in C (or a C array in C++) it will already be in the appropriate order and layout such that if the data is copied into a 1D array (or the pointer is passed as the result of `get_value_ptr`) it is already in the correct order/layout and will "just work".
+
+However, if you have a multi-dimensional array in Fortran or in Python/NumPy it may *not* be in the correct order. In Fortran, arrays are created in memory in "column major order"--however if you treat the first (left-most) index as the fastest changing, it is the same thing as C ordering where the C code would treat the last (right-most) index as the fastest changing. In Python, NumPy `ndarray`s are created as contiguous C-ordered blocks by default, but it is possible to create Fortran-ordered arrays, and if you take a view or slice of an array it is no longer a contiguous array and can't be passed without copying.
+
+### Proper ordering example
+
+The simplest way to explain the proper ordering is by example. Consider a `float` array with dimensions X = 4, Y = 3, and Z = 2. Such an array could be created and populated with the same values in the correct ordering and layout in the following ways:
+
+(See also https://www.visitusers.org/index.php?title=C_vs_Fortran_memory_order )
+
+C/C++
+```C
+float var[2][3][4];
+int x, y, z;
+float v = 0.0;
+for(z = 0; z < 2; z++)
+    for(y = 0; y < 3; y++)
+        for(x = 0; x < 4; x++)
+            var[z][y][x] = v += 0.01;
+```
+<!--
+float var_flat[24];
+memcpy(var_flat, var, 24*sizeof(float));
+printf("%f %f %f %f %f\n", var[0][0][0], var[0][0][1], var[0][0][2], var[0][1][0], var[1][0][0] );
+printf("%f %f %f %f %f\n", var_flat[0],  var_flat[1],  var_flat[2],  var_flat[4],  var_flat[12] );
+-->
+
+Fortran:
+<!--
+real, dimension(0:1,0:2,0:3) :: var
+integer:: x,y,z
+real:: v = 0.0
+zloop: do z = 0, 1
+   yloop: do y = 0, 2
+      xloop: do x = 0, 3
+        v = v + 0.01
+        var(z,y,x) = v ! Note the indices ordering here
+      end do xloop
+   end do yloop  
+end do zloop
+-->
+```Fortran
+real, dimension(0:3,0:2,0:1) :: var ! Note the reversal of the dimension sizes
+integer:: x,y,z
+real:: v = 0.0
+zloop: do z = 0, 1
+    yloop: do y = 0, 2
+        xloop: do x = 0, 3
+            v = v + 0.01
+            var(x,y,z) = v ! Note the indices ordering here
```

Comment:
`3, 2, 1` != `4, 3, 2` like I'd expect for equivalence to the other examples here

Reply:
Wait, what? It's been a while but I ended up running these examples a lot... you saw the `0:...` up in the `dimension(...)` call, right?

Reply:
Ah, I see now. I forgot that the way Fortran writes things ends up with closed intervals, not half-open intervals like C or Python.